### PR TITLE
Add capabilities to hide headers in the proxy directive

### PR DESCRIPTION
**Pull request by alexandreteles**
This patch tries to add capabilities to hide headers in the proxy directive, as requested on the #586 issue. Please review and let me know if I did anything wrong. Thank you for your time.

### CODE REVIEW

**Comment by mholt** [comment]
Thanks for the PR, Alexandre! I'm starting to take a look at it. If it does get accepted, it will need tests. Also, this looks like only the "hide" headers feature described in that issue, not the "ignore" headers feature, is that correct?

**Comment by alexandreteles** [comment]
Yes it's correct. It should be able to perform the same function with only one directive, the user just need to set what headers should be hidden. Nginx itself uses the response headers to determine how it should cache the proxy requests and the "ignore_headers" directive will tell what headers it should ignore on the cache process. As Caddy don't have a cache service (yet), the hide_headers will suffice.

File: middleware/proxy/upstream.go [hunk 1]
```diff
@@ -166,6 +169,12 @@ func parseBlock(c *parse.Dispenser, u *staticUpstream) error {
 			return c.ArgErr()
 		}
 		u.proxyHeaders.Add(header, value)
+	case "hide_header":
+		var header, value string
+		if !c.Args(&header, &value) {
+			return c.ArgErr()
+		}
+		u.hiddenHeaders.Add(header, value)
```
Comment (mholt) [inline middleware/proxy/upstream.go]: What is the purpose of storing the value here?

Comment (alexandreteles) [inline middleware/proxy/upstream.go]: There's no purpose. It was a simple copy/past from the proxy_header case above and I've forgotten to edit it. Sorry :smile_cat: 

If you have any other question, just tell me.

File: middleware/proxy/proxy.go [hunk 2]
```diff
@@ -107,8 +109,16 @@ func (p Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
 					}
 				}
 
+				var hiddenHeaders http.Header
+				if host.HiddenHeaders != nil {
+					hiddenHeaders = make(http.Header)
+					for header, _ := range host.HiddenHeaders {
+						hiddenHeaders.Add(header,"")
+					}
+				}
```
Comment (mholt) [inline middleware/proxy/proxy.go]: Maybe it's obvious, but why do we have to iterate and copy each header here into another http.Header type?

**Comment by mholt** [comment]
@alexandreteles What about combining your functionality into an existing syntax that mirrors the `header` directive, i.e.:

```text
proxy ... {
    proxy_header -Hide
    proxy_header X-Forwarded-For {host}
}
```

So this way, adding a header is the same, and hiding it is as easy as the leading `-`?

**Comment by mholt** [comment]
@alexandreteles How've things been?

Going to close this for now because of merge conflicts and inactivity, but when you're ready to re-visit this, feel free to re-open and rebase on master with your changes according to the comments so far. Definitely looking forward to these improvements. :smile:

**Pull request closed without merging**